src/data/finsum.lean

```diff
@@ -0,0 +1,657 @@
+There is work on `fincard` in the pipeline, which returns the cardinality of `X` if it
+is finite, and 0 otherwise.
+
+## Tags
+
+finsum, finsum_in, univ', finiteness
+-/
+
+open_locale classical
+
+universes u v w
+
+/-! # finset.univ' -- noncomputable univ (empty set if infinite) -/
+
+namespace finset
+
+/-- `univ' α` is the finset corresponding to all of α if α is finite, and the empty set otherwise.
+Note that it is noncomputable -/
+noncomputable def univ' (α : Type u) : finset α :=
+if h : nonempty (fintype α) then (classical.choice h).elems else ∅
+
+/-- `set.to_finset s` is the finset corresponding to `s` if `s` is finite, and the empty set
+    otherwise -/
+noncomputable def set.to_finset {α : Type u} (s : set α) : finset α :=
+if h : s.finite then h.to_finset else ∅
+
+variable {α : Type u}
+
+/-- `univ' α` is finset.univ if `α` is a fintype. -/
+lemma univ'_eq_univ [h : fintype α] : univ' α = univ :=
+by convert (dif_pos (nonempty.intro h))
+
+lemma univ'_eq_empty (h : ¬ nonempty (fintype α)) : univ' α = ∅ :=
+dif_neg h
+
+@[simp] lemma mem_univ' [h : fintype α] (x : α) : x ∈ univ' α :=
+(@univ'_eq_univ _ h).symm ▸ mem_univ _
+
+end finset
+
+/-! # noncomputable finite sums -/
+
+section finsum
+
+variables {α : Type u} {M : Type v} [add_comm_monoid M]
+
+/-- Sum of `f x` as `x` ranges over the elements of the support of `f`, if it's finite. Zero
+  otherwise. -/
+noncomputable def finsum (f : α → M) : M :=
+if h : (function.support f).finite then finset.sum (h.to_finset) f else 0
+
+/-- Sum of `f x` for `x ∈ s`, if `s ∩ f.support` is finite. Zero if it's infinite.  -/
+noncomputable def finsum_in (s : set α) (f : α → M) : M :=
+finsum (λ x, if x ∈ s then f x else 0)
+
+localized "notation `∑ᶠ` binders `, ` r:(scoped:67 f, finsum f) := r"
+  in big_operators
+
+localized "notation `∑ᶠ` binders ` in ` s `, ` r:(scoped:67 f, finsum_in s f) := r"
+  in big_operators
+
+open_locale big_operators
+
+lemma finsum_def (f : α → M) :
+  ∑ᶠ i : α, f i = if h : (function.support f).finite then finset.sum (h.to_finset) f else 0 := rfl
+
+lemma finsum_in_def (s : set α) (f : α → M) :
+  ∑ᶠ i in s, f i = finsum (λ x, if x ∈ s then f x else 0) := rfl
+
+-- function.support doesn't import data.finsupp and vice versa
+-- so the lemmas can't go in either place
+lemma function.support_eq_support {α : Type u} {β : Type v} [has_zero β] (f : α →₀ β) :
+  function.support f = f.support :=
+set.ext $ λ x, finsupp.mem_support_iff.symm
+
+namespace finsupp
+
+variables {β : Type v} [has_zero β]
+
+lemma support_finite (f : α →₀ β) : (function.support f).finite :=
+by { rw function.support_eq_support, exact f.support.finite_to_set }
+
+@[simp] lemma support_finite.to_finset (f : α →₀ β) : (support_finite f).to_finset = f.support :=
+finset.ext $ λ x, by simp
+
+/-- The natural `finsupp` induced by the function `f` given that it has finite support. -/
+noncomputable def of_support_finite
+  {f : α → β} (hf : (function.support f).finite) : α →₀ β :=
```

Comment:
`f` should be explicit here, because this is a def, not a theorem.

Reply:
Whoa, I didn't know there was a difference in the rules here!

Reply:
I am still unclear on what's going on here but I'll make the change.